lib/bech32/src/Codec/Binary/Bech32/Internal.hs

```diff
@@ -90,7 +95,45 @@ import qualified Data.ByteString as BS
 import qualified Data.ByteString.Char8 as B8
 
 {-------------------------------------------------------------------------------
-                          Human Readable Parts
+                                 Data Part
+-------------------------------------------------------------------------------}
+
+-- | Represents the data part of a Bech32 string, as defined here:
+--   https://git.io/fj8FS
+newtype DataPart = DataPart
+    { getDataPart :: [Word5] }
+    deriving Eq
+
+instance Monoid DataPart where
+    mempty = DataPart []
+
+instance Semigroup DataPart where
+    (DataPart d0) <> (DataPart d1) = DataPart $ d0 <> d1
+
+instance Show DataPart where
+    show (DataPart dp) = "DataPart " <> show ((word5ToChar Arr.!) <$> dp)
```

Comment:
I think we should put that one in the coding standards:

> Never define `Show` instances by hand

`Show` instances should really just output valid Haskell data-type, and we expect `read . show = id` at all time. If we need some human-friendly representation, then use `Buildable` instance from `fmt`, that's what the instance is for. Some preludes don't even allow programmers to write their own instance of `Show` as a guard.

We only have one exception which is our `ShowFmt` class we use to force QuickCheck to display counter examples using `Buildable` instances in order to get more readable counter examples.

Reply:
Hi @KtorZ!

Regarding:
> Never define `Show` instances by hand

I must admit I'm not yet convinced about this (though I could be).

My original reasoning for providing a `Show` instance here is for convenience. I can type the following within `ghci`:
```hs
> decode "abcdef1qpzry9x8gf2tvdw0s3jn54khce6mua7lmqqqxw"
Right (HumanReadablePart "abcdef", DataPart "qpzry9x8gf2tvdw0s3jn54khce6mua7l")
```
It's easy to verify (visually) that the human-readable and data parts correspond to the original string (with the separator and checksum removed). The data part is rendered with the Bech32 alphabet.

The standard derived instance is much less readable:
```hs
> decode "abcdef1qpzry9x8gf2tvdw0s3jn54khce6mua7lmqqqxw"
Right (HumanReadablePart "abcdef",DataPart {getDataPart = [Word5 {getWord5 = 0},Word5 {getWord5 = 1},Word5 {getWord5 = 2},Word5 {getWord5 = 3},Word5 {getWord5 = 4},Word5 {getWord5 = 5},Word5 {getWord5 = 6},Word5 {getWord5 = 7},Word5 {getWord5 = 8},Word5 {getWord5 = 9},Word5 {getWord5 = 10},Word5 {getWord5 = 11},Word5 {getWord5 = 12},Word5 {getWord5 = 13},Word5 {getWord5 = 14},Word5 {getWord5 = 15},Word5 {getWord5 = 16},Word5 {getWord5 = 17},Word5 {getWord5 = 18},Word5 {getWord5 = 19},Word5 {getWord5 = 20},Word5 {getWord5 = 21},Word5 {getWord5 = 22},Word5 {getWord5 = 23},Word5 {getWord5 = 24},Word5 {getWord5 = 25},Word5 {getWord5 = 26},Word5 {getWord5 = 27},Word5 {getWord5 = 28},Word5 {getWord5 = 29},Word5 {getWord5 = 30},Word5 {getWord5 = 31}]})
```
> we expect `read . show == id` at all times

That does make sense. I've added a corresponding `Read` instance, defined in terms of `dataPartFromText` (and a QuickCheck property to check for round trip compatibility.)

Thoughts?

Reply:
@KtorZ I've actually removed the hand-written `Read` and `Show` instances in a later commit.